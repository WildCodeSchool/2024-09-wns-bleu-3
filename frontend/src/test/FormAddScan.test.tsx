import "@testing-library/jest-dom";
// import userEvent from "@testing-library/user-event";
import { describe, expect, test, vi } from "vitest";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import ScanForm from "../components/FormAddScan";


//variable isDev qui vérifie si les tests s'exécutent dans un environnement de développement
const isDev = import.meta.env.DEV;

// createScanMock qui imite l'appel API pour créer un scan
const createScanMock = vi.fn(({ onCompleted }) => {
  if (onCompleted) onCompleted();
  return { loading: false };
});

// useCreateNewScanMutation : Renvoie la fonction simulée pour créer des scans
// useGetAllFrequencesQuery : Renvoie des données d'exemple pour les fréquences (options Quotidien et Hebdomadaire)
// useGetAllTagsQuery : Renvoie des données d'exemple pour les tags (le tag "Vitrine")
vi.mock("../generated/graphql-types", () => ({
  useCreateNewScanMutation: () => [createScanMock],
  useGetAllFrequencesQuery: () => ({
    data: {
      getAllFrequences: [
        { id: 1, name: "Daily", intervalMinutes: 60 },
        { id: 2, name: "Weekly", intervalMinutes: 30 },
      ],
    },
  }),
  useGetAllTagsQuery: () => ({
    data: { getAllTags: [{ id: 1, name: "Vitrine" }] },
  }),
}));

describe.runIf(isDev)("ScanForm", () => {

  // Ce test s'assure que le rendu du composant reste cohérent entre les exécutions de test
  test("matches the snapshot when rendered", () => {
    const { asFragment } = render(<ScanForm />);
    expect(asFragment()).toMatchSnapshot();
  })

  // Test de soumission du formulaire
  test("Create a scan", async () => {
    // const user = userEvent.setup();

    render(<ScanForm />);
    // screen.debug();

    const titleInput = screen.getByPlaceholderText("My Website Monitor");
    const urlInput = screen.getByPlaceholderText("https://example.com");
    const frequencySelect = screen.getByTestId('freqSelectButton');
    const tagSelect = screen.getByTestId('tagSelectButton');
    const submitButton = screen.getByRole("button");

    fireEvent.change(titleInput, { target: { value: "youtube" } });
    fireEvent.change(urlInput, { target: { value: "https://youtube.com" } });

    // Pour cibler le select de fréquence on cible le trigger
    fireEvent.click(frequencySelect);
    const vitrineOption = screen.getByRole("option", { name: "Daily" });
    fireEvent.click(vitrineOption);

    // Pareil pour tag, on cible le trigger pour faire apparaitre les options pour en sélectionner une
    fireEvent.click(tagSelect);
    const tagOption = screen.getByRole("option", { name: "Vitrine" });
    fireEvent.click(tagOption);


    // On délcenche la soumission du form
    fireEvent.click(submitButton);


    // Vérifie la soumission en contrôlant :
    // Que la mutation simulée a été appelée avec les bonnes données
    // Qu'elle a été appelée exactement une fois
    await waitFor(() => {
      expect(createScanMock).toHaveBeenCalledWith(
        expect.objectContaining({
          variables: {
            data: {
              title: "youtube",
              url: "https://youtube.com",
              tagIds: [1],
              frequencyId: 1,
            },
          },
        })
      );

      expect(createScanMock).toHaveBeenCalledTimes(1);

    });
  });

  //Ce test s'assure que le formulaire valide correctement les entrées utilisateur
  test("Display error message if title or url is empty", async () => {
    render(<ScanForm />);
    const submitButton = screen.getByRole("button");
    fireEvent.click(submitButton);
    await waitFor(() => {
      expect(screen.getByText("Title")).toBeInTheDocument();
      expect(screen.getByLabelText("URL to scan")).toBeInTheDocument();
    });
  });

});
