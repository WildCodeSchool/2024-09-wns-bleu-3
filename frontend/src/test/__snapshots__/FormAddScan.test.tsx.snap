// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScanForm > matches the snapshot when rendered 1`] = `
<DocumentFragment>
  <div
    class="max-w-md mx-auto bg-[#0a2540] rounded-xl shadow-lg border border-[#0c2d4d] p-6"
  >
    <h2
      class="text-xl font-semibold mb-4 text-white"
    >
      Start Scanning
    </h2>
    <form
      class="space-y-4"
    >
      <div
        class="space-y-2"
      >
        <label
          class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-sm font-medium text-gray-300"
          data-slot="label"
          for=":r0:-form-item"
        >
          Title
        </label>
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex h-9 w-full rounded-md border px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-neutral-950 placeholder:text-neutral-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:cursor-not-allowed disabled:opacity-50 md:text-sm dark:border-neutral-800 dark:file:text-neutral-50 dark:placeholder:text-neutral-400 dark:focus-visible:ring-neutral-300 mt-1 bg-[#0c2d4d] border-[#0e3359] text-white"
          id=":r0:-form-item"
          name="title"
          placeholder="My Website Monitor"
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-sm font-medium text-gray-300"
          data-slot="label"
          for=":r1:-form-item"
        >
          URL to scan
        </label>
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex h-9 w-full rounded-md border px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-neutral-950 placeholder:text-neutral-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:cursor-not-allowed disabled:opacity-50 md:text-sm dark:border-neutral-800 dark:file:text-neutral-50 dark:placeholder:text-neutral-400 dark:focus-visible:ring-neutral-300 mt-1 bg-[#0c2d4d] border-[#0e3359] text-white"
          id=":r1:-form-item"
          name="url"
          placeholder="https://example.com"
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-sm font-medium text-gray-300"
          data-slot="label"
          for="frequency"
        >
          Frequency
        </label>
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r3:"
          aria-expanded="false"
          class="data-[placeholder]:text-neutral-500 [&_svg:not([class*='text-'])]:text-neutral-500 focus-visible:border-neutral-950 focus-visible:ring-neutral-950/50 aria-invalid:ring-red-500/20 aria-invalid:border-red-500 flex h-9 items-center justify-between gap-2 rounded-md border px-3 py-2 text-sm whitespace-nowrap shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 *:data-[slot=select-value]:line-clamp-1 *:data-[slot=select-value]:flex *:data-[slot=select-value]:items-center *:data-[slot=select-value]:gap-2 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 dark:border-neutral-800 dark:data-[placeholder]:text-neutral-400 dark:[&_svg:not([class*='text-'])]:text-neutral-400 dark:focus-visible:border-neutral-300 dark:focus-visible:ring-neutral-300/50 dark:aria-invalid:ring-red-900/20 dark:dark:aria-invalid:ring-red-900/40 dark:aria-invalid:border-red-900 w-full mt-1 bg-[#0c2d4d] border-[#0e3359] text-white"
          data-placeholder=""
          data-slot="select-trigger"
          data-state="closed"
          data-testid="freqSelectButton"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            data-slot="select-value"
            style="pointer-events: none;"
          >
            Select a frequency
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down size-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
        <select
          aria-hidden="true"
          name="frequency"
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          tabindex="-1"
        >
          <option
            value="1"
          >
            Daily
          </option>
          <option
            value="2"
          >
            Weekly
          </option>
        </select>
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-sm font-medium text-gray-300"
          data-slot="label"
          for=":r4:-form-item"
        >
          Tag
        </label>
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r5:"
          aria-expanded="false"
          class="data-[placeholder]:text-neutral-500 [&_svg:not([class*='text-'])]:text-neutral-500 focus-visible:border-neutral-950 focus-visible:ring-neutral-950/50 aria-invalid:ring-red-500/20 aria-invalid:border-red-500 flex h-9 items-center justify-between gap-2 rounded-md border px-3 py-2 text-sm whitespace-nowrap shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 *:data-[slot=select-value]:line-clamp-1 *:data-[slot=select-value]:flex *:data-[slot=select-value]:items-center *:data-[slot=select-value]:gap-2 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 dark:border-neutral-800 dark:data-[placeholder]:text-neutral-400 dark:[&_svg:not([class*='text-'])]:text-neutral-400 dark:focus-visible:border-neutral-300 dark:focus-visible:ring-neutral-300/50 dark:aria-invalid:ring-red-900/20 dark:dark:aria-invalid:ring-red-900/40 dark:aria-invalid:border-red-900 w-full mt-1 bg-[#0c2d4d] border-[#0e3359] text-white"
          data-placeholder=""
          data-slot="select-trigger"
          data-state="closed"
          data-testid="tagSelectButton"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            data-slot="select-value"
            style="pointer-events: none;"
          >
            Select a tag
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down size-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
        <select
          aria-hidden="true"
          name="tag"
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          tabindex="-1"
        >
          <option
            value=""
          />
          <option
            value="1"
          >
            Vitrine
          </option>
        </select>
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 bg-blue-400 text-white hover:bg-blue-500 h-9 px-4 py-2 w-full"
        type="submit"
      >
        Start Scanning
      </button>
    </form>
  </div>
</DocumentFragment>
`;
